### PR TITLE
Enable ssh keys authentication on Azure platform.

### PR DESCRIPTION
**Pull request by LiliDeng**
This change enables ssh key authentication on Azure platform.
Pass public key like this -SSHPublicKey D:\lili\test.pub - this is used for deploy new azure deployment
Pass private key like this -SSHPrivateKey D:\lili\test.ppk - this is used for authentication when copy files and execute command against VM.

### CODE REVIEW

File: Libraries/TestHelpers.psm1 [hunk 1]
```diff
@@ -58,12 +58,15 @@ function Upload-RemoteFile($uploadTo, $port, $file, $username, $password, $usePr
 	if (!$maxRetry) {
 		$maxRetry = 10
 	}
-
+	if ($sshPublicKey -and $sshPrivateKey) {
+		$usePrivateKey = $true
+		$sshKey = $sshPrivateKey
+	}
 	while($retry -le $maxRetry) {
 		# TODO: $UsePrivateKey is not enabled yet
```
Comment (lizzha) [inline Libraries/TestHelpers.psm1]: remove the TODO?

File: TestControllers/AzureController.psm1 [hunk 2]
```diff
@@ -165,6 +167,7 @@ Class AzureController : TestController
 
 		# Used in CAPTURE-VHD-BEFORE-TEST.ps1 and some cases
 		Set-Variable -Name ARMImageName -Value $this.ARMImageName -Scope Global -Force
+		Set-Variable -Name SSHPrivateKey -Value $this.SSHPrivateKey -Scope Global -Force
```
Comment (lizzha) [inline TestControllers/AzureController.psm1]: it's already set in TestController

File: Run-LisaV2.ps1 [hunk 3]
```diff
@@ -86,6 +86,7 @@ Param(
 	# [Optional] Parameters for changing framework behavior.
 	[int]    $TestIterations = 1,
 	[string] $XMLSecretFile = "",
+	[string] $SSHPrivateKey = "",
```
Comment (lizzha) [inline Run-LisaV2.ps1:89]: could you help me understand why SSHPrivateKey is in the parameter, and SSHPublicKey is in the secret file?
Also can we add document about how to generate them?

File: Libraries/TestHelpers.psm1 [hunk 4]
```diff
@@ -58,12 +58,15 @@ function Upload-RemoteFile($uploadTo, $port, $file, $username, $password, $usePr
 	if (!$maxRetry) {
 		$maxRetry = 10
 	}
-
+	if ($sshPublicKey -and $sshPrivateKey) {
```
Comment (lizzha) [inline Libraries/TestHelpers.psm1]: do they come from global variable? can we make it more clear by adding $global: ?

**Review by lizzha — CHANGES_REQUESTED** [review]

File: Run-LisaV2.ps1 (hunk 3 above)
Comment (LiliDeng) [reply Run-LisaV2.ps1:89]: SSHPublicKey - it is a string I set in XMLSecretFile file, pass it into json template when deploy VM.
SSHPrivateKey - it is a file, suffix is .ppk. also it is a string. But I didn't find a way to convert this sting to the private key file. Thoughts?

**Comment by santoshx** [comment]
May be title should be "Enable ssh 'keys' authentication on Azure platform" .

File: Run-LisaV2.ps1 (hunk 3 above)
Comment (LiliDeng) [reply Run-LisaV2.ps1:89]: @lizzha I update code to pass SSHPrivateKey and SSHPublicKey into the command line, let know any concern.

File: TestControllers/AzureController.psm1 [hunk 5]
```diff
@@ -87,15 +87,26 @@ Class AzureController : TestController
...
+
+		if ($this.SSHPrivateKey) {
+			Write-LogDbg "Use SSH key authentication. Reset password as empty."
+			$this.VmPassword = ""
+			if (!$this.UseExistingRG -and !$this.SSHPublicKey) {
+				throw "Please set -SSHPublicKey and -SSHPrivateKey at the same time for a new deployment."
+			}
+		}
+		if (!$this.SSHPrivateKey -and !$this.VmPassword) {
+			throw "Please set -SSHPrivateKey or linuxTestPassword."
```
Comment (LiliDeng) [inline TestControllers/AzureController.psm1]: @lizzha This will cause a UnitTests_Pester error, does it make sense?

**Review by santoshx — COMMENTED** [review]
Please add new options information to how-to-use.md. Otherwise looks good.

File: Libraries/Azure.psm1 [hunk 6]
```diff
@@ -1004,7 +1004,7 @@ Function Generate-AzureDeployJSONFile ($RGName, $ImageName, $osVHD, $RGXMLData,
     #$PublicIPv6Name = $($RGName.ToUpper() -replace '[^a-z]') + "PublicIPv6"
     $PublicIPv6Name = "PublicIPv6"
     $sshPath = '/home/' + $user + '/.ssh/authorized_keys'
-    $sshKeyData = ""
+    $sshKeyData = $global:sshPublicKey
```
Comment (juhlee-microsoft) [inline Libraries/Azure.psm1:1007]: I assume this key-based access change is for Azure only. Then should we document somewhere for public such as how to add the key in the local system, and how to share the key?

File: Libraries/TestHelpers.psm1 [hunk 7]
```diff
@@ -58,12 +58,14 @@ function Upload-RemoteFile($uploadTo, $port, $file, $username, $password, $usePr
 	if (!$maxRetry) {
 		$maxRetry = 10
 	}
-
+	if ($global:sshPrivateKey) {
+		$usePrivateKey = $true
```
Comment (juhlee-microsoft) [inline Libraries/TestHelpers.psm1:62]: This variable should have initial value.

File: Run-LisaV2.ps1 (hunk 3 above)
Comment (juhlee-microsoft) [reply Run-LisaV2.ps1:89]: Yep, I think we need to add contexts about how to use in the document.

File: TestControllers/AzureController.psm1 [hunk 8]
```diff
@@ -87,15 +87,25 @@ Class AzureController : TestController
 			$secrets = $this.XMLSecrets.secrets
 			$azureConfig.Subscription.SubscriptionID = $secrets.SubscriptionID
 			$azureConfig.TestCredentials.LinuxUsername = $secrets.linuxTestUsername
-			$azureConfig.TestCredentials.LinuxPassword = $secrets.linuxTestPassword
-			$azureConfig.ResultsDatabase.server = $secrets.DatabaseServer
-			$azureConfig.ResultsDatabase.user = $secrets.DatabaseUser
-			$azureConfig.ResultsDatabase.password = $secrets.DatabasePassword
-			$azureConfig.ResultsDatabase.dbname = $secrets.DatabaseName
+			$azureConfig.TestCredentials.LinuxPassword = if ($secrets.linuxTestPassword) { $secrets.linuxTestPassword } else { "" }
+			$azureConfig.ResultsDatabase.server = if ($secrets.DatabaseServer) { $secrets.DatabaseServer } else { "" }
```
Comment (juhlee-microsoft) [inline TestControllers/AzureController.psm1:91]: Those 4 line changes will need to follow up the rest of the code change. If I understand correct, this should allow the secret xml file contains none of DB information. So we should add 'filtering' lines in each SQL statement in the LISAv2. It's not about this PR, but recommend to add a new issue for future followup.

**Review by juhlee-microsoft — CHANGES_REQUESTED** [review]

File: Libraries/TestHelpers.psm1 (hunk 7 above)
Comment (LiliDeng) [reply Libraries/TestHelpers.psm1:62]: it is the parameter of function Upload-RemoteFile

**Comment by juhlee-microsoft** [comment]
Unit test result should be fixed 

Describing Test if module AzureController PrepareTestEnvironment is valid
    [-] Should prepare test env 236ms
      Expected no exception to be thrown, but an exception "Please set -SSHPrivateKey or linuxTestPassword." was thrown from D:\_work\1\s\TestControllers\AzureController.psm1:107 char:4
          +             throw "Please set -SSHPrivateKey or linuxTestPassword."
          +             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
      108: 		{ $azureController.PrepareTestEnvironment($fakeXML) } | Should Not Throw
      at <ScriptBlock>, D:\_work\1\s\UnitTests\PowerShell\TestControllers\AzureController.Tests.ps1: line 108

**Review by juhlee-microsoft — APPROVED** [review]
Let's wait @lizzha's sign-off also

**Comment by LiliDeng** [comment]
@lizzha Could you please review again and let know any comments?

**Review by lizzha — APPROVED** [review]

**Pull request merged**